src/Stratis.Bitcoin/Consensus/BestChainIndexer.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Consensus
+{
+    /// <summary>
+    /// An indexer of the consensus chain.
+    /// </summary>
+    public class BestChainIndexer
+    {
+        private readonly IConsensusManager consensusManager;
+        private readonly Dictionary<int, ChainedHeader> blocksByHeight = new Dictionary<int, ChainedHeader>();
+        private readonly object lockObj = new object();
+
+        internal BestChainIndexer(IConsensusManager consensusManager)
+        {
+            this.consensusManager = consensusManager;
+        }
+
+        public ChainedHeader Tip => this.consensusManager.Tip;
+
+        internal void Initialize(ChainedHeader chainedHeader)
```

Comment:
I think it will be a good idea to make `blocksByHeight` not readonly and create it at the initialization providing height of the tip as capacity value to the constructor of the dictionary

Reply:
I was thinking of that, why do you think its better? if the chain indexer will be its own component (not strongly bound to CM) then yes I would agree.
I wonder if this is what we should do actually, but then the add/remove blocks will no be internal (unless we could create an internal interface hmm...)

Reply:
>why do you think its better? 

It will preallocate enough space to fit all the items without resizing the underlying array.
Basically it's just more efficient in terms of resources usage

Reply:
I agree about the initial sizing of the dictionary, sizing to tip size will prevent the resizing of blocksByHeight during initialization.
basically when you reach the full capacity it resizes (and if I recall right it resize using the size of the prime number higher than current items count, anyway the source is here https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Collections/Generic/Dictionary.cs)

It may still happen during node lifecycle tho, but allocating the right size as startup is a good thing.

removing readonly doesn't impact on external visibility, it will still be private, you just can now instantiate it outside of constructor.
(if in the constructor we had already the tip value, than we could have left it readonly and instantiated in the constructor itself but I don't think it's the case)

Reply:
This is very minor IMO it only effects the perf on startup and this dictionary anyway grows, I don't mind though.